File: haskell/rot26.hs
-- Haskell rot26 implementation
-- by ArmoredStone
-- 24.01.2025

import System.IO
import System.Environment (getArgs)

main :: IO ()
main = do
    hSetEncoding stdout utf8
    hSetEncoding stdin utf8
    
    args <- getArgs
    let mode = head (head args)
        textChars = unwords (tail args)
        result = modeCase mode textChars
    putStrLn (result)

modeCase :: Char -> [Char] -> [Char]
modeCase mode text = case mode of
    'e' -> map applySucc26times text
    'd' -> map applyPred26times text
    _ -> "Invalid mode, use e or d"

applySucc26times :: Char -> Char
applySucc26times = (!! 26) . iterate succ

applyPred26times :: Char -> Char
applyPred26times = (!! 26) . iterate pred

File: go/rot26.go
// Go rot26 implementation
// by ArmoredStone
// 24.01.2025

package main

import (
	"fmt"
	"os"
	"strings"
)

func rotateChar(char rune, mode string) rune {
	shift := 26 
	if mode == "left" {
		return rune(int(char) - shift)
	} else if mode == "right" {
		return rune(int(char) + shift)
	} else {
		fmt.Println("Invalid mode. Use 'left' or 'right'.")
		return rune(0)
	}
}

func rotateString(str string, mode string) string {
	var result strings.Builder
	for _, char := range str {
		result.WriteRune(rotateChar(char, mode))
	}
	return result.String()
}

func main() {
	if len(os.Args) != 3 {
		fmt.Println("Usage: go run main.go <left/right> <string>")
		return
	}

	mode := os.Args[1]
	inputString := os.Args[2]
	result := rotateString(inputString, mode)
	fmt.Println(result)
}
